lib/compiler.js

```diff
@@ -29,36 +79,97 @@ Compiler.prototype.generate = function generate(code, options) {
-    ') {',
-    getCode(code, options.runtimeLint) + ';',
-    '});',
-    'api.exportApply(exports);',
-    '/// -------------------------------------',
-    '/// ------ BEM-XJST User-code End -------',
-    '/// -------------------------------------\n'
-  ].join('\n');
+    'var ' + exportName + ';',
+    '(function(global) {',
+      'function buildBemXjst(libs) {',
+        'var exports;',
+
+        '/* BEM-XJST Runtime Start */',
+        'var ' + exportName + ' = function(module, exports) {',
+           bundles[this.engineName] + ';',
+          'return module.exports || exports.' + exportName + ';',
+        '}({}, {});',
+
+        'var api = new ' + exportName + '(' + JSON.stringify(options) + ');',
+
+        '/* BEM-XJST User-code Start */',
+        'api.compile(function(' +
+          require('./bemxjst').prototype.locals.join(', ') +
+        ') {',
+          getCode(code, options.runtimeLint) + ';' +
+        '});',
+
+        'exports = api.exportApply(exports);',
+        'if (libs) exports.BEMContext.prototype._libs = libs;',
+
+        'return exports;',
+      '};',
+
+      options.commonJSModules,
+
+      'var glob = this.window || this.global || this;',
+      'var exp = typeof exports !== "undefined" ? exports : global;',
+
+      // Provide with CommonJS
+      deps.commonJSNames.length === 0 || (
+        'if (typeof module==="object" && typeof module.exports==="object") {' +
+        'exp["' + exportName + '"] = buildBemXjst();' +
+        'exp["' + exportName + '"].libs = {};' + (
+          deps.commonJSNames.map(function(dep) {
+            return 'exp["' + exportName + '"].libs["' + dep + '"] = ' +
+              'glob && typeof glob["' + dep + '"] ' +
+              '!== "undefined" ? ' +
+              'glob["' + dep + '"] : require("' + dep + '");';
+          })
+        ) + '}'),
+
+      // Provide to YModules
+     'if (typeof modules === "object") {',
+       'modules.define("' + exportName + '",' +
+         JSON.stringify(deps.ym) + ',' +
+         'function(provide' + deps.ymVars + ') { ' +
+           'var engine = buildBemXjst();' +
+           deps.ymLibs +
+           'provide(engine);' +
+         '}' +
+       ');',
+     '}',
+
+      // Provide to global scope
+      deps.globalNames.length !== 0 ? (
+        exportName + ' = buildBemXjst(glob);' +
+        'exp["' + exportName + '"] = ' + exportName + ';' +
+        'exp["' + exportName + '"].libs = {};' +
+        deps.commonJSNames.map(function(dep) {
+          return 'typeof glob["' + dep + '"] !== "undefined" && (' +
+            'exp["' + exportName + '"].libs["' + dep + '"] = ' +
+            'glob["' + dep + '"]);';
+        }).join('')
+      ) : (
+        exportName + '= buildBemXjst(this.global);' +
+        'global["' + exportName + '"] = ' + exportName + ';'
+      ),
+
+    '})(typeof window !== "undefined" ? ' +
+      'window : typeof global !== "undefined" ? global : this);'
```

Comment:
I feel like this code smell, but I don't atm know how to bake it better.

I'd say it's fine solution since compile should generate code with exports.